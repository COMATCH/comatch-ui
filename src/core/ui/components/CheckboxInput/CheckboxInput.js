import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faCheck } from '@fortawesome/free-solid-svg-icons/faCheck';
import noop from 'lodash/noop';
import { InputLabel } from '../InputLabel';
import { InputError } from '../InputError';

import './CheckboxInput.scss';

const propTypes = {
    display: PropTypes.oneOf(['block', 'inline']),
    id: PropTypes.string,
    inputError: PropTypes.string,
    label: PropTypes.oneOfType([PropTypes.node, PropTypes.string]),
    name: PropTypes.string.isRequired,
    onChange: PropTypes.func,
    checked: PropTypes.oneOfType([PropTypes.bool, PropTypes.string]),
    value: PropTypes.oneOfType([PropTypes.bool, PropTypes.string]),
};

const defaultProps = {
    display: 'block',
    id: null,
    inputError: '',
    label: '',
    onChange: noop,
    checked: null,
    value: '',
};

export const CheckboxInput = ({ display, id, inputError, label, onChange, name, checked, value }) => {
    const classes = classNames('Input', 'CheckboxInput', display, { 'has-error': inputError });

    return (
        <div id={id} className={classes}>
            <label>
                <input type="checkbox" name={name} onChange={onChange} checked={checked} value={value} />
                <div className="CheckboxInput__input" role="checkbox">
                    <FontAwesomeIcon className="CheckboxInput__tick" icon={faCheck} />
                </div>
                {label && <InputLabel className="CheckboxInput__label" span={true} text={label} />}
            </label>
            {inputError && <InputError text={inputError} />}
        </div>
    );
};

CheckboxInput.defaultProps = defaultProps;

CheckboxInput.propTypes = propTypes;
