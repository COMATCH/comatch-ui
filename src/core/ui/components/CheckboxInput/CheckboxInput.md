
## CheckboxInput

From [`CheckboxInput.js`](CheckboxInput.js)



prop | type | default | required | description
---- | :----: | :-------: | :--------: | -----------
**checked** | `Union<Boolean \| String>` | `null` | :x: | 
**display** | `Enum('block','inline')` | `'block'` | :x: | 
**id** | `String` | `null` | :x: | 
**inputError** | `String` | `''` | :x: | 
**label** | `Union<ReactNode \| String>` | `''` | :x: | 
**name** | `String` |  | :white_check_mark: | 
**onChange** | `Function` | `noop` | :x: | 
**value** | `Union<Boolean \| String>` | `''` | :x: | 



