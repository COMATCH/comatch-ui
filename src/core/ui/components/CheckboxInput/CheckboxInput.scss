@import '../../../styles/variables.scss';

.CheckboxInput {
    display: inline-block;

    &.block {
        display: block;
        text-align: left;
    }

    // Whitespace
    &:not(.block) ~ &:not(.block) {
        margin-left: 30px;
    }

    // Common styles
    &__input,
    &__label {
        display: inline-block;
        vertical-align: middle;
    }

    &__label,
    &__label * {
        line-height: $co-line-height-sm;
    }

    // Actual input
    input {
        display: none; // hide

        // When actually input has state :checked
        // make the tick of the pseudo-checkbox show up
        &:checked ~ .CheckboxInput__input {
            .CheckboxInput__tick {
                transform: scale(1) translate3d(-50%, -50%, 0);
            }
        }
    }

    // Pseudo input
    &__input {
        // coloring
        background-color: $co-color-white;
        border: 1px solid $co-color-mid-gray;
        border-radius: 1px;
        // positioning
        position: absolute;
        top: 3px;
        left: 0;
        // sizing
        width: 15px;
        height: 15px;

        // Tick of pseudo checkbox
        .CheckboxInput__tick {
            color: $co-color-primary;
            // positioning
            position: absolute;
            top: 50%;
            left: 50%;
            // animation
            transition: transform 400ms cubic-bezier(0.45, 1.8, 0.5, 0.75);
            transform: scale(0) translate3d(-50%, -50%, 0);
            transform-origin: top left;
        }
    }

    &__label {
        color: $co-color-black;
        font-weight: normal;
        font-size: $co-font-sm;
        line-height: $co-line-height-sm;
    }

    &__input ~ &__label {
        margin-left: 10px;
    }

    label {
        cursor: pointer;
        position: relative;
        padding-left: 15px;
    }

    // Whitepsace
    & ~ & {
        margin-top: 5px;
    }

    &.inline {
        .InputLabel {
            margin-bottom: 0;
        }
    }
}
