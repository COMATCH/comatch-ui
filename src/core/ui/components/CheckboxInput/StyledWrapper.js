import React, { forwardRef } from 'react';
import styled from 'styled-components';
import { palette, typography } from '../../../styles/variables';

export const generateDisplayStyling = ({ display }) => {
    switch (display) {
        case 'block':
            return `
                display: block;
                text-align: left;
            `;

        case 'inline':
            return 'display: inline-block;';

        default:
            return '';
    }
};

export const StyledCheckboxInput = styled.div`
    background-color: ${palette.white};
    border: 1px solid ${palette.midGray};
    border-radius: 1px;
    height: 15px;
    margin-top: 2px;
    position: relative;
    width: 15px;

    // Tick of pseudo checkbox
    .CheckboxInput__tick {
        color: ${palette.primary};
        // positioning
        position: absolute;
        top: 50%;
        left: 50%;
        // animation
        transition: transform 400ms cubic-bezier(0.45, 1.8, 0.5, 0.75);
        transform: scale(0) translate3d(-50%, -50%, 0);
        transform-origin: top left;
    }
`;

export const StyledWrapper = styled(forwardRef(({ display, ...rest }, ref) => <div ref={ref} {...rest} />))`
    ${generateDisplayStyling}
    & ~ & {
        margin-top: 5px;
    }

    // Whitespace
    &:not(.block) ~ &:not(.block) {
        margin-left: 30px;
    }

    input:not(.ToggleSwitch) {
        display: none;

        // When actually input has state :checked
        // make the tick of the pseudo-checkbox show up
        &:checked ~ ${StyledCheckboxInput} .CheckboxInput__tick {
            transform: scale(1) translate3d(-50%, -50%, 0);
        }
    }

    label {
        cursor: pointer;
        position: relative;
        display: inline-flex;
    }

    ${StyledCheckboxInput} {
        vertical-align: middle;
    }

    .CheckboxInput__label {
        vertical-align: top;
    }

    ${StyledCheckboxInput},
    .CheckboxInput__label {
        display: inline-block;
    }

    .CheckboxInput__label {
        color: ${palette.black};
        display: inline-block;
        font-weight: normal;
        font-size: ${typography.fontSm};
        flex: 1;
        line-height: ${typography.lineHeightSm};
        margin-bottom: 0;

        & * {
            line-height: ${typography.lineHeightSm};
        }
    }

    ${StyledCheckboxInput} ~ .CheckboxInput__label {
        margin-left: 10px;
    }
`;

export const StyledToggleSwitchInput = styled.main`
    .ToggleSwitch {
        width: 30px;
        height: 15px;
        background-color: #dce2e2;
        color: #0095b3;
        border-radius: 7.5px;
        appearance: none;
        margin: 5px 15px 0 10px;
    }

    .ToggleSwitch:before,
    .ToggleSwitch:after {
        content: '';
        display: block;
    }

    .ToggleSwitch:before {
        background-color: #ffffff;
        border-radius: 50%;
        width: 25px;
        height: 25px;
        margin: -5px 0 0 -15px;
        box-shadow: 0 1px 2px 0 rgba(62, 62, 62, 0.2);
        border: solid 1px #dce2e2;
    }

    .ToggleSwitch:checked {
        background-color: rgba(0, 149, 179, 0.3);
    }

    .ToggleSwitch:checked:before {
        background-color: currentColor;
        transform: translateX(125%);
        border: none;
        margin: -5px 0 0 -15px;
    }

    .ToggleSwitch:focus {
        outline: 0;
    }
`;
