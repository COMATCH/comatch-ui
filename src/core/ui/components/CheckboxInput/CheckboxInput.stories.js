import React, { Component } from 'react';
import Highlight from 'react-highlight';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { CheckboxInput } from './CheckboxInput';

storiesOf('CheckboxInput', module)
    .add('single no label', () => (
        <>
            <CheckboxInput name="isFoo" />
            <Highlight className="html">{'<CheckboxInput name="isFoo" />'}</Highlight>
        </>
    ))
    .add('single with label', () => (
        <>
            <CheckboxInput label="Foo" name="isFoo" />
            <Highlight className="html">{'<CheckboxInput label="Foo" name="isFoo" />'}</Highlight>
        </>
    ))
    .add('multiple with label', () => (
        <div>
            <CheckboxInput label="Foo" name="isFoo" value="1" />
            <CheckboxInput label="Bar" name="isBar" value="2" />
        </div>
    ))
    .add('multiple with label and display inline', () => (
        <div>
            <CheckboxInput label="Foo" name="isFoo" display="inline" />
            <CheckboxInput label="Bar" name="isBar" value="2" display="inline" />
        </div>
    ))
    .add('multiple with label, display inline and display block', () => (
        <div>
            <CheckboxInput label="Foo" name="isFoo" />
            <CheckboxInput label="Bar" name="isBar" />
            <CheckboxInput label="Atomic" name="isAtomic" value="3" display="inline" />
            <CheckboxInput label="Radioactive" name="isRadioactive" value="4" display="inline" />
            <CheckboxInput label="Supersonic" name="isSupersonic" value="5" display="inline" />
            <CheckboxInput label="Hey!" name="isHey" value="6" />
            <CheckboxInput label="Ho!" name="isHo" value="7" />
        </div>
    ))
    .add('single with label and inputError', () => (
        <>
            <CheckboxInput label="Foo" name="isFoo" inputError="Alarm! Meltdown Imminent!" />
            <Highlight className="html">
                {'<CheckboxInput label="Foo" name="isFoo" inputError="Alarm! Meltdown Imminent!" />'}
            </Highlight>
        </>
    ))
    .add('multiple with label and inputError', () => (
        <div>
            <CheckboxInput label="Foo" name="isFoo" inputError="Alarm! Meltdown Imminent!" />
            <CheckboxInput label="Bar" name="isBar" inputError="Alarm! Meltdown Imminent!" />
            <CheckboxInput label="Atomic" name="isAtomic" inputError="Alarm! Meltdown Imminent!" />
            <CheckboxInput label="Radioactive" name="isRadioactive" />
            <CheckboxInput label="Supersonic" name="isSupersonic" />
        </div>
    ))
    .add('multiple with label, name, value and onChange', () => {
        const handleChange = (event) => action('changed')(event.target.name, event.target.checked);
        return (
            <div>
                <CheckboxInput label="Foo" name="isFoo" value="1" onChange={handleChange} />
                <CheckboxInput label="Bar" name="isBar" value="2" onChange={handleChange} />
                <CheckboxInput label="Atomic" name="isAtomic" value="3" onChange={handleChange} />
            </div>
        );
    })
    .add('in a container', () => <CheckboxInputContainer />);

class CheckboxInputContainer extends Component {
    state = { available: false };

    handleChange = (event) => {
        this.setState({ available: event.target.checked });
        action('changed in container')(event.target.name, event.target.checked);
    };

    render() {
        return (
            <>
                <CheckboxInput label="Available" name="available" onChange={this.handleChange} />
                <Highlight className="html">
                    {'<CheckboxInput label="Available" name="available" onChange={function() {}} />'}
                </Highlight>
            </>
        );
    }
}
