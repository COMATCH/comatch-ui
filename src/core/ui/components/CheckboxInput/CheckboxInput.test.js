import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import noop from 'lodash/noop';
import { CheckboxInput } from './CheckboxInput';
import '../../../config/tests/setup';

describe('CheckboxInput', () => {
    it('should render correctly', () => {
        const component = renderer.create(<CheckboxInput name="" />).toJSON();
        expect(component).toMatchSnapshot();
    });

    it('should have the correct default props', () => {
        const wrapper = mount(<CheckboxInput name="" />);
        expect(wrapper.prop('display')).toEqual('block');
        expect(wrapper.prop('id')).toEqual(null);
        expect(wrapper.prop('inputError')).toEqual('');
        expect(wrapper.prop('label')).toEqual('');
        expect(wrapper.prop('onChange')).toEqual(noop);
        expect(wrapper.prop('checked')).toEqual(null);
        expect(wrapper.prop('value')).toEqual('');
    });

    it('should have the correct passed props', () => {
        const mockFunction = jest.fn();
        const props = {
            display: 'inline',
            id: 'test id',
            inputError: 'test inputError',
            label: 'test label',
            name: 'test name',
            onChange: mockFunction,
            checked: true,
            value: 'test value',
        };

        const wrapper = mount(<CheckboxInput {...props} />);
        expect(wrapper.prop('display')).toEqual('inline');
        expect(wrapper.prop('id')).toEqual('test id');
        expect(wrapper.prop('inputError')).toEqual('test inputError');
        expect(wrapper.prop('label')).toEqual('test label');
        expect(wrapper.prop('name')).toEqual('test name');
        expect(wrapper.prop('onChange')).toEqual(mockFunction);
        expect(wrapper.prop('checked')).toEqual(true);
        expect(wrapper.prop('value')).toEqual('test value');
    });

    it('should render the label', () => {
        const props = {
            label: 'test label',
            name: 'test name',
        };

        const wrapper = mount(<CheckboxInput {...props} />);
        expect(wrapper.find('label').text()).toEqual('test label');
    });

    it('should render the input error', () => {
        const props = {
            inputError: 'test inputError',
            name: 'test name',
        };

        const wrapper = mount(<CheckboxInput {...props} />);
        expect(wrapper.find('InputError').text()).toEqual('test inputError');
    });

    it('should pass name, onChange, checked and value props to input element', () => {
        const mockFunction = jest.fn();
        const props = {
            name: 'test name',
            onChange: mockFunction,
            checked: true,
            value: 'test value',
        };

        const wrapper = mount(<CheckboxInput {...props} />);
        expect(wrapper.find('input').prop('name')).toEqual('test name');
        expect(wrapper.find('input').prop('onChange')).toEqual(mockFunction);
        expect(wrapper.find('input').prop('checked')).toEqual(true);
        expect(wrapper.find('input').prop('value')).toEqual('test value');
    });

    it('should run the onClick function', () => {
        const mockFunction = jest.fn();
        const props = {
            name: 'test name',
            onChange: mockFunction,
            checked: true,
            value: 'test value',
        };

        const wrapper = mount(<CheckboxInput {...props} />);
        wrapper.find('label').simulate('click');
        expect(mockFunction.mock.calls.length).toEqual(0);
    });
});
