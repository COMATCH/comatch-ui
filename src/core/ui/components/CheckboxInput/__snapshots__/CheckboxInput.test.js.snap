// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxInput should render correctly 1`] = `
<div
  className="sc-bxivhb bJyOBc Input CheckboxInput block"
  id={null}
>
  <label>
    <input
      checked={null}
      name=""
      onChange={[Function]}
      type="checkbox"
      value=""
    />
    <div
      className="sc-htpNat jpWaUF CheckboxInput__input"
      role="checkbox"
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-check fa-w-16 CheckboxInput__tick"
        data-icon="check"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </div>
    
  </label>
  
</div>
`;
